### ISSUE TITLE
Repeating text changing doesn't work.

### ISSUE THREAD

**Issue by gerardroche** [NeoVintageous/NeoVintageous#10]
<a href="https://github.com/jordwalke"><img src="https://avatars0.githubusercontent.com/u/977348?v=3" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [jordwalke](https://github.com/jordwalke)**
_Sunday Mar 17, 2013 at 07:33 GMT_
_Originally opened as https://github.com/guillermooo/Vintageous/issues/148_

----

Type the following text:

```
abcdefg
abcdefg
```

Visually highlight abc and then type "c" for change.

Then replace abc with 123 and hit escape.

You will now have

```
123defg
abcdefg
```

Then position your cursor on the next "a".

Hit repeat (".") - it will not replace abc with 123.

Instead, you will have:

```
123defg
123abcdefg
```

**Comment by gerardroche** [comment]
<a href="https://github.com/guillermooo"><img src="https://avatars1.githubusercontent.com/u/684948?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [guillermooo](https://github.com/guillermooo)**
_Sunday Mar 17, 2013 at 09:51 GMT_

----

I've been postponing this one for too long, but it looks like it's going to be rather tricky to get working. I seems surprisingly simple in Vintage, but I can't find a way to make it work in Vintageous.

**Comment by gerardroche** [comment]
<a href="https://github.com/jordwalke"><img src="https://avatars0.githubusercontent.com/u/977348?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [jordwalke](https://github.com/jordwalke)**
_Sunday Mar 17, 2013 at 09:58 GMT_

----

I haven't thought about it in too much depth, but couldn't it be as simple as replaying the last set of keystrokes that resulted in a text edit? Or one segment of operations that were glued together in the undo stack?

**Comment by gerardroche** [comment]
<a href="https://github.com/guillermooo"><img src="https://avatars1.githubusercontent.com/u/684948?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [guillermooo](https://github.com/guillermooo)**
_Sunday Mar 17, 2013 at 10:27 GMT_

----

It is indeed as simple as that, but, as far as I can tell, some intermediary steps are getting in the way and polluting the undo stack. Separating genuine edits to the buffer from these other bookkeeping commands is what's not so obvious here. Most likely, I will have to revert `vi_enter_insert_mode` and friends to stand-alone commands, and stop processing them through the usual `set_command` path that legit edit commands must use.

Confusingly, though, the marking and gluing of edits is working for insert mode, and I don't think entering insert mode is too much different from entering visual mode right now.

**Comment by gerardroche** [comment]
<a href="https://github.com/guillermooo"><img src="https://avatars1.githubusercontent.com/u/684948?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [guillermooo](https://github.com/guillermooo)**
_Sunday Mar 17, 2013 at 10:28 GMT_

----

Also, a quick experiment to enable grouping of edits in replace mode failed as well. Must take a closer look.

**Comment by gerardroche** [comment]
<a href="https://github.com/guillermooo"><img src="https://avatars1.githubusercontent.com/u/684948?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [guillermooo](https://github.com/guillermooo)**
_Sunday Mar 17, 2013 at 11:59 GMT_

----

I've found a fix, at least a partial one, but the downside is I don't fully understand why it's working :/

**Comment by gerardroche** [comment]
<a href="https://github.com/jordwalke"><img src="https://avatars0.githubusercontent.com/u/977348?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [jordwalke](https://github.com/jordwalke)**
_Sunday Mar 17, 2013 at 20:14 GMT_

----

I'd be curious to see a gist of what ended up working - it might be a good way for me to get a good overview of the Vintageous codebase. Maybe you could point out your initial reasoning.

**Comment by gerardroche** [comment]
<a href="https://github.com/guillermooo"><img src="https://avatars1.githubusercontent.com/u/684948?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [guillermooo](https://github.com/guillermooo)**
_Monday Mar 18, 2013 at 02:40 GMT_

----

I'm trying to figure out when is it ok to call the `mark_undo...` and `glue...` S3 built-in commands. Vintage does a bit of extra work with them, and I'm not taking exactly the same approach in Vintageous. It remains to be seen whether Vintage's way is the only valid one.

In the meantime, my experiments show that you cannot call `glue_marked_undo_groups` from a command that is meant to be subsumed into the glued group.

That would explain why I can make `v3lcXXX<ESC>` work, but not `v2ed` (unpublished code). In the first case, the gluing occurs after you press `ESC`, which is a completely separate step, so everything is fine. In the second case, the `run.py` code transitions to the next mode while the full run of the command hasn't finished. In fact, if I override the code dealing with the `follow_up_mode` in `run.py`, `v2ed` can be repeated too just as good, but then, of course, the mode handling gets broken.

I think the solution requires to move `.do_follow_up_mode()` from `run.py` to `state.py`'s `.run()` and put it after the full evaluation of the vi command is done (the call to `vi_run`). I seem to remeber I've tried this before quite unsuccessfully, but I'll have to revisit the idea.

There's also another piece in the puzzle. For me, `v3lcXXX<ESC>` started to be repeatable by simply replacing `mark_undo_groups_for_gluing` for `maybe_mark_undo_groups_for_gluing` in the current code. The former should be used in some places where a whole new 'group' is desired (like by pressing `v`), but if I do use it, things stop working.

**Comment by gerardroche** [comment]
<a href="https://github.com/guillermooo"><img src="https://avatars1.githubusercontent.com/u/684948?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [guillermooo](https://github.com/guillermooo)**
_Monday Mar 18, 2013 at 03:06 GMT_

----

Might be related, because I used to handle the next mode transition in `state.py`'s `.reset()`. #149.

**Comment by gerardroche** [comment]
<a href="https://github.com/jordwalke"><img src="https://avatars0.githubusercontent.com/u/977348?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [jordwalke](https://github.com/jordwalke)**
_Monday Mar 18, 2013 at 05:06 GMT_

----

Here's an issue from Vintage, where I had some difficulties with gluing and redo (not sure if the same troubles would come with repeat as well).

https://github.com/sublimehq/Vintage/issues/153

It may not be related at all, but at least I can share in your confusion. This one issue related to yank was the Vintage deal-breaker bug IMHO.

**Comment by gerardroche** [comment]
<a href="https://github.com/guillermooo"><img src="https://avatars1.githubusercontent.com/u/684948?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [guillermooo](https://github.com/guillermooo)**
_Monday Mar 18, 2013 at 05:41 GMT_

----

I'm almost there, at least on the surface it looks like it... It seems GH won't let me create a fork of my own repo, and because some problems with hg-git (I use Mercurial), I cannot create temp branches either without breaking my local repo.

Basically, the mode change cycle needs to be defined in more stricter terms. I'm taking a first step toward that goal, but it won't be finished with this patch when it's ready. Shoehorning the mode handling code into S3's macro and repeat restrictions feels too forcible at the moment. It isn't too bad now, but it isn't too clear either.

**Comment by gerardroche** [comment]
<a href="https://github.com/guillermooo"><img src="https://avatars1.githubusercontent.com/u/684948?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [guillermooo](https://github.com/guillermooo)**
_Monday Mar 18, 2013 at 05:46 GMT_

----

(Your referenced issue is definitely related, but will probably have to be dealt with separately after this one.)

**Comment by gerardroche** [comment]
<a href="https://github.com/jordwalke"><img src="https://avatars0.githubusercontent.com/u/977348?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [jordwalke](https://github.com/jordwalke)**
_Monday Mar 18, 2013 at 07:29 GMT_

----

As far as I can tell, my referenced issue works excellently in Vintageous so that's great news.

**Comment by gerardroche** [comment]
<a href="https://github.com/guillermooo"><img src="https://avatars1.githubusercontent.com/u/684948?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [guillermooo](https://github.com/guillermooo)**
_Tuesday Mar 19, 2013 at 05:15 GMT_

----

@jordow Can we close this one?

**Comment by gerardroche** [comment]
<a href="https://github.com/jordwalke"><img src="https://avatars0.githubusercontent.com/u/977348?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [jordwalke](https://github.com/jordwalke)**
_Tuesday Mar 19, 2013 at 07:32 GMT_

----

It doesn't seem to work. In fact, `ce` used to work correctly, but now that doesn't even work (I downloaded master and tried just a moment ago).

**Comment by gerardroche** [comment]
<a href="https://github.com/jordwalke"><img src="https://avatars0.githubusercontent.com/u/977348?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [jordwalke](https://github.com/jordwalke)**
_Tuesday Mar 19, 2013 at 07:36 GMT_

----

Here's the repro:

```
test1

test2

test3
```

Then in normal mode, do `cw` on "test1" and type anything like "hello".

Then, go to the next line and hit the repeat key. Nothing happens.

**Comment by gerardroche** [comment]
<a href="https://github.com/guillermooo"><img src="https://avatars1.githubusercontent.com/u/684948?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [guillermooo](https://github.com/guillermooo)**
_Tuesday Mar 19, 2013 at 08:20 GMT_

----

Weird. I made sure to test on my end and it was working. I'll take a look at this tonight.

**Comment by gerardroche** [comment]
<a href="https://github.com/guillermooo"><img src="https://avatars1.githubusercontent.com/u/684948?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [guillermooo](https://github.com/guillermooo)**
_Tuesday Mar 19, 2013 at 08:26 GMT_

----

I get the same. It was definitely working before. I must have messed up something while refactoring.

**Comment by gerardroche** [comment]
<a href="https://github.com/guillermooo"><img src="https://avatars1.githubusercontent.com/u/684948?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [guillermooo](https://github.com/guillermooo)**
_Tuesday Mar 19, 2013 at 08:30 GMT_

----

ce in normal mode is working fine here, though.

**Comment by gerardroche** [comment]
<a href="https://github.com/guillermooo"><img src="https://avatars1.githubusercontent.com/u/684948?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [guillermooo](https://github.com/guillermooo)**
_Tuesday Mar 19, 2013 at 08:36 GMT_

----

Are you ussing S3 dev 3022? I've just upgraded and everything is back to normal.

**Comment by gerardroche** [comment]
<a href="https://github.com/guillermooo"><img src="https://avatars1.githubusercontent.com/u/684948?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [guillermooo](https://github.com/guillermooo)**
_Tuesday Mar 19, 2013 at 08:54 GMT_

----

Ok, so ce in normal mode works. ce<esc>, then "." works as well. What isn't working is ceXXX then ".".

**Comment by gerardroche** [comment]
<a href="https://github.com/jordwalke"><img src="https://avatars0.githubusercontent.com/u/977348?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [jordwalke](https://github.com/jordwalke)**
_Tuesday Mar 19, 2013 at 09:27 GMT_

----

I was using 3021. I can update and then try again, but in 3022 are both `ce` and `cw` working for you?

**Comment by gerardroche** [comment]
<a href="https://github.com/guillermooo"><img src="https://avatars1.githubusercontent.com/u/684948?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [guillermooo](https://github.com/guillermooo)**
_Tuesday Mar 19, 2013 at 10:05 GMT_

----

They are, but they don't repeat correctly.

**Comment by gerardroche** [comment]
<a href="https://github.com/guillermooo"><img src="https://avatars1.githubusercontent.com/u/684948?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [guillermooo](https://github.com/guillermooo)**
_Tuesday Mar 19, 2013 at 10:33 GMT_

----

BTW, updating is a requirement, because the repeat command "." depends on this change introduced in 3022:

> API: view.command_history(0, True) now returns the last modifying command, as expected

**Comment by gerardroche** [comment]
<a href="https://github.com/jordwalke"><img src="https://avatars0.githubusercontent.com/u/977348?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [jordwalke](https://github.com/jordwalke)**
_Wednesday Mar 20, 2013 at 04:47 GMT_

----

I see. Then what version were you running when everything worked correctly for you (including repeat?)
